microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/properties/CompletionData.java

```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
+* Copyright (c) 2023 Red Hat Inc. and others.
+*
+* This program and the accompanying materials are made available under the
+* terms of the Eclipse Public License v. 2.0 which is available at
+* http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+* which is available at https://www.apache.org/licenses/LICENSE-2.0.
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lsp4mp.services.properties;
+
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4mp.commons.utils.JSONUtility;
+
+import com.google.gson.JsonObject;
+
+/**
+ * Represents data sent to for completionItem/resolve
+ *
+ * { 'data': { 'propertyName': {...}, 'uri': {...} } }
+ */
+public class CompletionData {
+
+	private static final String DATA_PROPERTY_NAME_FIELD = "propertyName";
+
+	private static final String DATA_URI_FIELD = "uri";
+
+	private String propertyName;
+
+	private String uri;
+
+	public String getPropertyName() {
+		return propertyName;
+	}
+
+	public void setPropertyName(String propertyName) {
+		this.propertyName = propertyName;
+	}
+
+	public String getUri() {
+		return uri;
+	}
+
+	public void setUri(String uri) {
+		this.uri = uri;
+	}
+
+	public static CompletionData getCompletionData(CompletionItem unresolved) {
+		return JSONUtility.toModel(unresolved.getData(), CompletionData.class);
+	}
+
+	/**
+	 * Returns the minimal data to resolve a completion item.
+	 *
+	 * @param propertyName the property name of the completion item.
+	 * @param uri          the uri of the text document.
+	 * @return the minimal data to resolve a completion item.
+	 */
+	public static JsonObject createCompletionData(String propertyName, String uri) {
```

Comment:
You should be able to set `data` directly as an instance of `CompletionData`. It will be automatically turned into JSON when the message is serialized.

Here's my suggestion:
- add a constructor for `CompletionData` that accepts `propertyName` and `uri`, and use this instead of `createCompletionData(...)`
- add a no-argument constructor that sets `propertyName` and `uri` to `null` (this is needed for the JSON serialization to work)